vent/coordinator/coordinator.py

```diff
@@ -3,65 +3,114 @@
 from typing import List, Dict
 
 import vent
-from vent.common.message import SensorValues, ControlSetting, Alarm, IPCMessageCommand
+from vent.common.message import SensorValues, ControlSetting, Alarm
 from vent.common.values import ValueName
+from vent.common import values
 from vent.common.message import SensorValueNew
-from vent.controller.control_module import get_control_module
-from vent.coordinator.ipc import IPC
+import vent.controller.control_module
+from vent.coordinator.ipc import IPC, IPCMessage, IPCCommand
 from vent.coordinator.process_manager import ProcessManager
 
 
 class CoordinatorBase:
     def __init__(self, sim_mode=False):
         # get_ui_control_module handles single_process flag
         # TODO: SHARED_ is a better prefix than COPY_, as not all fields are copy
-        self.control_module = get_control_module(sim_mode)
         self.COPY_sensor_values = None
         self.COPY_alarms = None
         self.COPY_active_alarms = {}
         self.COPY_logged_alarms = []
         self.COPY_control_settings = {}
         self.COPY_tentative_control_settings = {}
         self.COPY_last_message_timestamp = None
-        
-    def get_sensors(self) -> Dict[ValueName, SensorValueNew]:
-        # returns SensorValues struct
-        pass
+        self.lock = threading.Lock()
+        self._is_running = threading.Event()
 
-    # def get_alarms(self) -> List[Alarm]:
-    #     # returns list of Alarm structs
-    #     pass
+    def get_sensors(self) -> Dict[ValueName, SensorValueNew]:
+        with self.lock:
+            sensor_values = self.COPY_sensor_values
+        if sensor_values is None:
+            return None
+        # TODO: merge the API such that controller and coordinator use the same SensorValue class
+        else:
+            res = {
+                ValueName.PIP: SensorValueNew(ValueName.PIP, sensor_values.pip, sensor_values.timestamp,
```

Comment:
Do we still have "loop_counter" as part of sensor values? i.e. to indicate that the controller is still making actual progress.

Reply:
The controller process can be detected using both rpc timeout and loop_counter. But I prefer we not deal with it in this pull request.